pkg/k8s/service.go

```diff
@@ -483,6 +483,89 @@ func NewClusterService(id ServiceID, k8sService *Service, k8sEndpoints *Endpoint
 	return svc
 }
 
+// ParseClusterService parses a ClusterService and returns a Service
+func ParseClusterService(svc *serviceStore.ClusterService) *Service {
+	var ip net.IP
+	var ipStr string
+	ports := serviceStore.PortConfiguration{}
+	for ipStr, ports = range svc.Frontends {
+		ip = net.ParseIP(ipStr)
+		break
+	}
+	svcInfo := &Service{
+		FrontendIP:      ip,
+		IsHeadless:      false,
+		IncludeExternal: true,
+		Shared:          true,
+		TrafficPolicy:   loadbalancer.SVCTrafficPolicyCluster,
+		Ports:           map[loadbalancer.FEPortName]*loadbalancer.L4Addr{},
+		Labels:          svc.Labels,
+		Selector:        svc.Selector,
+		Type:            loadbalancer.SVCTypeClusterIP,
+	}
+
+	for name, port := range ports {
+		p := loadbalancer.NewL4Addr(loadbalancer.L4Type(port.Protocol), uint16(port.Port))
+		portName := loadbalancer.FEPortName(name)
+		if _, ok := svcInfo.Ports[portName]; !ok {
+			svcInfo.Ports[portName] = p
+		}
+	}
+
+	return svcInfo
+}
+
+// DeepEquals returns true if both services are equal
+func (s *Service) EqualsClusterService(svc *serviceStore.ClusterService) bool {
+	switch {
+	case (s == nil) != (svc == nil):
+		return false
+	case (s == nil) && (svc == nil):
+		return true
+	}
+
+	var ip net.IP
+	var ipStr string
+	ports := serviceStore.PortConfiguration{}
+	for ipStr, ports = range svc.Frontends {
+		ip = net.ParseIP(ipStr)
+		break
+	}
+
+	if s.FrontendIP.Equal(ip) &&
+		s.IsHeadless == false &&
+		s.IncludeExternal == true &&
+		s.Shared == true &&
+		s.TrafficPolicy == loadbalancer.SVCTrafficPolicyCluster &&
+		s.HealthCheckNodePort == 0 &&
+		len(s.NodePorts) == 0 &&
+		len(s.K8sExternalIPs) == 0 &&
+		len(s.LoadBalancerIPs) == 0 &&
+		len(s.LoadBalancerSourceRanges) == 0 &&
+		comparator.MapStringEquals(s.Labels, svc.Labels) &&
+		comparator.MapStringEquals(s.Selector, svc.Selector) &&
+		s.SessionAffinity == false &&
+		s.SessionAffinityTimeoutSec == 0 &&
+		s.Type == loadbalancer.SVCTypeClusterIP {
```

Comment:
I think we should move this to a helper function. It's quite a long list of conditionals that could use some explanation via moving to a helper.

Reply:
The whole function is an "Equals" function. I'll add a comment to refer to the parse function right above for the detail.

Reply:
Can we have this function auto generated please?

Reply:
This can not be auto-generated, as it compares two different types. We need this to avoid re-creating each service when the backends change, as the ClusterService in kvstore represents both the service and the backends in the same value.